trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/plugins/generator/cacsl2boogietranslator/Boogie2ACSL.java

```diff
@@ -89,6 +95,85 @@ public BacktranslatedExpression translateExpression(
 	}
 
+	/**
+	 * Provides a specialized translation for the 'ensures' clause in a procedure contract. Specifically, it adds
+	 * conjuncts of the form 'x == \old(x)' for all global variables that are not modifiable by the procedure, and that
+	 * are off-heap (because we assume conservatively that the procedure may modify the heap).
+	 *
+	 * @param expression
+	 *            The original 'ensures' clause to translate
+	 * @param context
+	 *            The context, i.e., the location of the C function definition
+	 * @param modifiableGlobals
+	 *            The set of variables listed in the Boogie procedure "modifies" clause
+	 * @return A backtranslated ensures clause with the additional conjuncts
+	 */
+	public BacktranslatedExpression translateEnsuresExpression(
+			final de.uni_freiburg.informatik.ultimate.boogie.ast.Expression expression, final ILocation context,
+			final Set<de.uni_freiburg.informatik.ultimate.boogie.ast.IdentifierExpression> modifiableGlobals) {
+		final var intType = new CPrimitive(CPrimitives.INT);
+		final var boolRange = BigInterval.booleanRange();
+
+		if (expression == null) {
+			return computeOldVarEqualities(modifiableGlobals)
+					.map(e -> new BacktranslatedExpression(e, intType, boolRange)).orElse(null);
+		}
+
+		final var mainExpression = translateExpression(expression, context);
+		final var oldVarEqualities = computeOldVarEqualities(modifiableGlobals);
+		if (oldVarEqualities.isPresent()) {
+			return new BacktranslatedExpression(
+					new BinaryExpression(Operator.LOGICAND, mainExpression.getExpression(), oldVarEqualities.get()),
+					new CPrimitive(CPrimitives.INT), BigInterval.booleanRange());
+		}
+
+		return mainExpression;
+	}
+
+	private Optional<Expression> computeOldVarEqualities(
+			final Set<de.uni_freiburg.informatik.ultimate.boogie.ast.IdentifierExpression> modifiableGlobals) {
+		final var modifiableNames = modifiableGlobals.stream().map(e -> e.getIdentifier()).collect(Collectors.toSet());
+
+		return mSymbolTable.getGlobalScope().entrySet().stream()
+				.flatMap(e -> makeOldVarEquality(e.getKey(), e.getValue(), modifiableNames).stream())
+				.reduce((eq1, eq2) -> new BinaryExpression(Operator.LOGICAND, eq1, eq2));
+	}
+
+	private Optional<Expression> makeOldVarEquality(final String cId, final SymbolTableValue stv,
+			final Set<String> modifiableNames) {
+		final boolean isTypeDecl = stv.getBoogieDecl() instanceof TypeDeclaration;
+		if (isTypeDecl) {
+			// Type declarations are not global variables
+			return Optional.empty();
+		}
+
+		final var boogieName = stv.getBoogieName();
+		if (modifiableNames.contains(boogieName)) {
+			// No equalities for modifiable variables.
+			return Optional.empty();
+		}
+
+		if (!mMapping.hasVar(boogieName, stv.getDeclarationInformation())) {
+			// We omit the conjunct if the variable cannot be backtranslated and output a warning (sound but imprecise).
+			mReporter.accept("Unknown variable: " + boogieName);
```

Comment:
Can and should we check here, if this variable is just some auxiliary variable (`#length`, `#valid`, `#race`) and don't report a warning in that case?

Reply:
We don't have such a check for the normal backtranslation of expressions, right?

I expect that this check is mostly triggered by such variables.  The point of the warning is to indicate potential problems, e.g. for inductiveness of the contract. Unless we can soundly rule out that the contract becomes non-inductive due to some specific variable, I think the warning is justified.